common/djangoapps/third_party_auth/middleware.py

```diff
@@ -23,3 +25,23 @@ def get_redirect_uri(self, request, exception):
             redirect_uri = pipeline.AUTH_DISPATCH_URLS[auth_entry]
 
         return redirect_uri
+
+
+class PipelineQuarantineMiddleware(object):
+
+    def process_view(self, request, view_func, view_args, view_kwargs):  # pylint: disable=unused-argument
+        """
+        Check the session to see if we've quarantined the user to a particular
+        step of the authentication pipeline; if so, look up which modules the
+        user is allowed to browse to without breaking the pipeline. If the view
+        that's been requested is outside those modules, then flush the session.
+        """
+        running_pipeline = request.session.get('partial_pipeline')
+
+        if not running_pipeline:
+            return
+
+        view_module = view_func.__module__
+        quarantined_modules = request.session.get('sso_quarantined_modules', None)
```

Comment:
Should be called `third_party_auth_quarantined_modules` - "SSO" is too ambiguous. Also would be nice if `third_party_auth` provided a method to set/unset this, so other apps don't have to manipulate the session directly nor hard-code this string.